### PR TITLE
fix(scan): locate the URL anywhere in a pipeline.md line — 5 of 6 documented shapes defeated both dedupe gates

### PR DESCRIPTION
**Pull request by bravely**
## Problem

Both `data/pipeline.md` dedupe gates assume the URL is the first cell after the checkbox:

```js
// gate 1 — loadSeenUrls, scan.mjs:1047
/- \[[ x]\] (https?:\/\/\S+)/

// gate 2 — collectSeenCompanyRoles, scan.mjs:1388
/^- \[[ x]\]\s+\S+\s*\|\s*([^|\n]+?)\s*\|\s*([^|\n]+?)\s*(?:\|[^\n]*)?$/
```

That holds for exactly one line shape — the one `appendToPipeline` writes. The modes document five others that lead with a report number, a markdown link, or a strikethrough. Gate 1 misses those. Gate 2's leading `\S+` then eats the first cell instead of a URL, shifting both captures one column left, so it does not merely fail — it adds a **wrong** key.

| Line shape (source) | gate 1 | gate 2 key (before) |
|---|---|---|
| `- [ ] {url} \| {co} \| {role}` — `appendToPipeline` | ✅ | `acme::staff engineer` |
| `- [x] #143 \| {url} \| …` — `modes/pipeline.md:38,63` | miss | `https://…/jobs/2::acme` |
| `- [x] [144](reports/…) \| {url} \| …` — `reconcile-pipeline.mjs` | miss | `https://…/jobs/3::acme` |
| `- [x] #-- \| {url} \| skipped (…)` — `modes/pipeline.md:23,35` | miss | `https://…/jobs/4::skipped pre screen…` |
| `- [x] ~~{url} \| {co} \| {role}~~ — expired` — `modes/pipeline.md:13` | miss | `acme::backend engineer posting expired…` |
| `- [x] ~~{co} \| {role}~~ — expired` — `modes/oferta.md:15`, `auto-pipeline.md:29` | miss (no URL) | none |

Row 2 is verbatim from the worked example in `modes/pipeline.md:63-64`.

Measured on a real `data/pipeline.md` from a few months of ordinary use: **54% of URL-bearing lines are invisible to gate 1**, and **43% of gate-2 matches put the URL in the company slot**. Dozens of the missed URLs are in neither `scan-history.tsv` nor `applications.md`, so nothing dedupes them — those are hand-pasted and ingest-path entries, which never pass through a scanner and so never reach scan history. They are precisely what this gate exists to cover.

## Root cause

The comment above gate 2 documents one shape (`"- [ ] {url} | {company} | {title}"`) — the one the scanner writes. Both gates read the file unscoped, so they see all six.

This bug class was already fixed **inside this same function**. Three lines above gate 2, the `applications.md` branch reads:

```js
// applications.md — header-aware parse (tracker-parse.mjs, #954). The old
// positional regex captured the wrong cells on customized layouts (e.g. with a
// Location column), so the seen-set keyed on garbage and dedup misfired.
```

The pipeline branch below still does exactly what #954 removed.

## The fix

Locate the URL wherever it sits in the line rather than assuming column 1, and take company/role from the cells after it.

Seeding is deliberately **asymmetric**, mirroring the rule the scan-history branch already applies:

- Live entries (pending, processed) seed both gates.
- **Expired entries (`~~…~~`) seed the URL gate only.** Seeding a dead posting's role key "would let a dead SF URL bury a live NY req" (the function's own docstring); the scan-history branch enforces this with `if (status !== 'added') continue;`, and strikethrough is the pipeline's expression of that same state.
- **Pre-screen discards seed the URL gate only** — there is no company/role pair in that shape.

A fix that parsed all six shapes uniformly would introduce the dead-URL bug, so the tests pin the asymmetry too.

## Tests

All six documented shapes are now pinned in `tests/scan-company-role-dedup.test.mjs` and `tests/scan-url-dedup.test.mjs`, which previously covered only the shape the scanner writes — which is how five documented shapes drifted out of support unnoticed. Cases are sourced from the mode docs, so the tests fail if the documented format changes.

## Related

#2065 fixed the *value* half of this same symptom (URL string variation) in these loaders; this is the format half. #2382 is an open perf refactor of the same three loaders and will touch these lines.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved pipeline deduplication across varied checkbox, report-link, and strikethrough formats.
  * Excluded non-HTTP, expired, pre-screen discard, and malformed entries from duplicate detection.
  * Preserved URL normalization, TTL behavior, and application-tracker parsing.
  * Ensured roles spanning multiple cities are consolidated into a single pipeline entry.

* **Tests**
  * Added comprehensive coverage for supported pipeline formats, URL extraction, company-role matching, and end-to-end deduplication.

### CODE REVIEW

File: scan.mjs [hunk 1]
```diff
@@ -1026,6 +1026,80 @@ export function normalizeUrlForDedup(url) {
...
+ * The leading checkbox marker of a `data/pipeline.md` entry.
+ *
+ * Only ` ` and `x` are recognized, matching the pre-existing gates: `- [!]`
+ * marks a URL that could not be fetched, which is not evidence a posting was
+ * surfaced.
+ */
+const PIPELINE_CHECKBOX_RE = /- \[[ x]\]\s+/;
+
+/** As `PIPELINE_CHECKBOX_RE`, anchored — used where the old gate anchored too. */
+const PIPELINE_CHECKBOX_ANCHORED_RE = /^- \[[ x]\]\s+/;
```
[the remark this answers is not in the document]
Comment (bravely) [reply scan.mjs]: Good catch — this is a real false positive, fixed in 96e8440, though with `^\s*` rather than the suggested `^`.

The hazard is worse than "prose that happens to contain checkbox syntax": it is a widening relative to the gate this replaces. The old regex required the URL to sit *immediately* after the checkbox, so `Reminder: - [ ] chase the recruiter, posting is https://…` never matched. Mine matched the checkbox anywhere and then the URL anywhere, so it did — seeding a live posting's URL and silently burying it. A false dedupe hides a real job, which is the costlier direction of error here.

I went with `^\s*` because the literal `^` overcorrects in two ways:

| line | old gate | this PR (before) | `^` | `^\s*` |
|---|---|---|---|---|
| `Reminder: - [ ] … https://…/9 (still open)` | — | **seeds** | — | — |
| `  - [ ] https://…/10 \| Acme Corp \| Staff Engineer` | seeds | seeds | **—** | seeds |

Indented entries deduped under the old adjacency-based regex, so `^` would drop postings that are being deduped today. It would also make `PIPELINE_CHECKBOX_RE` byte-identical to the constant declared right below it.

`^\s*` closes the hole with no narrowing. Both edges are now pinned in `tests/scan-url-dedup.test.mjs`.

The company/role gate keeps its stricter `^- \[` anchor, so I renamed that constant `PIPELINE_CHECKBOX_ANCHORED_RE` → `PIPELINE_CHECKBOX_STRICT_RE` — "anchored" no longer distinguishes the two, since the difference is now only indentation. Widening that one would seed *more* role keys, and a single role key suppresses every posting sharing it.

**Review by slegarraga — APPROVED** [review]
Solid fix. The six documented shapes are covered, and `^\s*` is the right balance: it closes the inline-prose false positive while preserving indented entries. The asymmetric seeding for expired and pre-screen rows matches the scan-history rule. Looks good to me.

**Comment by santifer** [comment]
The diagnosis is right and the asymmetry argument is the correct one, @bravely. I probed `main` directly and it is worse than "shapes that fail to dedupe":

- `- [x] #143 | https://x/2 | Acme Corp | AI PM | 4.2/5 | PDF ✅` seeds the key `https://x/2::acme corp`, with the URL sitting in the company slot.
- `- [x] ~~https://x/9 | Acme Corp | SRE~~ — posting expired` seeds `acme corp::sre posting expired liveness sweep`, a dead-role key the scan-history branch takes explicit care never to create.

Your PR fixes all four shapes and I verified the six documented ones against `modes/pipeline.md`, `modes/oferta.md`, `modes/auto-pipeline.md` and `reconcile-pipeline.mjs` rather than taking the list on faith. The tests are the opposite of decoration too: spawning real child processes because `scan.mjs` resolves paths at import time, asserting exact set size instead of membership.

**One root cause, three symptoms, and it lands on the shape that already worked.**

```js
PIPELINE_URL_RE = /https?:\/\/[^\s|)~]+/
```

`~` is an RFC 3986 unreserved character, legal anywhere in a URL. Banning it from the class to strip the strikethrough wrapper truncates real URLs:

```console
$ node -e "const RE=/https?:\/\/[^\s|)~]+/; for (const s of ['https://x.com/~acme/jobs/2','https://x.com/jobs/3(eu)']) console.log(JSON.stringify(s), '->', JSON.stringify(s.match(RE)?.[0]))"
"https://x.com/~acme/jobs/2" -> "https://x.com/"
"https://x.com/jobs/3(eu)" -> "https://x.com/jobs/3(eu"
```

Tilde paths are not exotic: `~user` URLs are still everywhere on university and older corporate hosts, and plenty of ATS links carry parenthesised region suffixes.

By your own standard this is the expensive direction. `https://x.com/~acme/jobs/2` collapsing to `https://x.com/` means (a) that entry stops deduping, which is the duplicate-resurfacing this PR exists to kill, and (b) a bare-origin key enters the seen-set, so anything else on that host can now false-dedupe against it.

Third symptom, same cause: the whole-line `line.includes('~~')` guard makes a live entry lose its pair when a documented `note:` column happens to contain a strikethrough.

```
- [ ] https://x/4 | Acme Corp | Data Engineer | note: replaces ~~old req~~
main → ["acme corp::data engineer"]
PR   → (none)
```

Scoping the strikethrough to the **entry wrapper** rather than treating `~` as a global signal fixes all three, and it is roughly two lines: match the wrapper at the entry boundary instead of excluding the character everywhere.

Rest of it holds: SHA matches, merge clean against today's `main`, full suite `3414 passed, 0 failed, 1 warnings` (the warning is the pre-existing one).

Push the wrapper-scoped version and this goes in. It touches `scan.mjs` and `data/pipeline.md` parsing, both of which the web reads, so I will coordinate that side when it lands. Nothing for you to do there.

**Comment by bravely** [comment]
You're right, and the framing is the useful part: one root cause, and it lands on the shape that already worked. I was treating `~` as a signal rather than as a character, so the wrapper leaked into two places it had no business being.

Fixed in 5a5df81. Expiry is now read at the **entry boundary** — the wrapper only counts when it opens the entry body, and the body is unwrapped once so every caller downstream parses the same cell sequence either way. An unclosed `~~` still reads as expired, since the opening marker is what carries the meaning.

That let the URL class drop both exclusions:

```js
const PIPELINE_URL_RE = /https?:\/\/[^\s|]+/;
```

Whitespace and `|` only. `|` cannot appear unencoded in a URL and is this format's cell separator, so it is the one boundary that is safe by construction rather than by luck — which is exactly what `~` and `)` were not.

All three symptoms, against the merged branch:

| input | before | after |
|---|---|---|
| `- [ ] https://x.com/~acme/jobs/2 \| Acme Corp \| SRE` | `https://x.com/` | `https://x.com/~acme/jobs/2` |
| `- [ ] https://x.com/jobs/3(eu) \| Acme Corp \| Data Eng` | `https://x.com/jobs/3(eu` | `https://x.com/jobs/3(eu)` |
| `- [ ] https://x/4 \| Acme Corp \| Data Engineer \| note: replaces ~~old req~~` | no pair | `acme corp::data engineer` |

The asymmetry is unchanged and still pinned: expired and pre-screen entries seed a URL key and no role key.

Also merged `main` in, which needed a real resolution — #2382 landed and split `loadSeenUrls` into `collectSeenUrls(sources, policy)` plus a filesystem wrapper, so the pipeline branch moved into the pure collector and walks `pipelineText` rather than reading the file. Happy side effect: the URL tests no longer need the child-process harness, since the thing under test now takes text directly. The company/role suite keeps its end-to-end run.

New coverage for all three cases above, plus the two boundary cases from the earlier round. Full suite `3446 passed, 0 failed, 0 warnings`.

One thing I did **not** touch, flagging rather than folding in: the `applications.md` branch of `collectSeenUrls` still uses `[^\s|)]+`, so a parenthesised URL keys differently depending on which source it came from. That predates this PR and the markdown-link reasoning for excluding `)` is at least arguable there, so it seemed like yours to call rather than mine to bundle.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Comment by santifer** [comment]
The round-1 ask is fully closed, @bravely, and I verified it by execution rather than by reading: `~acme/jobs/11` and `jobs/12(eu)` both survive intact now, the strikethrough regex is anchored, and the entry with `note: replaces ~~old req~~` keeps its pair. All six documented forms produce the right keys where the baseline produced one correct key out of five. That is the fix working.

Two regressions the tests do not cover, both landing on shapes that already worked, which is the exact thing this PR argues against.

**D1 — `local:` entries lose their company/role key.** `scan.mjs:1230` bails when no cell matches the URL regex:

```
"- [ ] local:jds/linkedin-pm-ai.md | Acme Corp | Staff Engineer"
   baseline gate 2 → ["acme corp::staff engineer"]
   patched  gate 2 → []
```

The old `\S+` matched `local:…` without meaning to. So the docstring at `scan.mjs:1150-1151` ("`local:` entries are deliberately not matched — the gates this feeds have always been http(s)-only") is true for gate 1 and **false for gate 2**. `local:` is documented at `modes/pipeline.md:123`, and a preexisting comment in the very file you edit (`tests/scan-url-dedup.test.mjs:67-69`) names this exact failure: *"or a local JD would be re-added on every scan"*. Cheap direction (a duplicate reappears, nothing hides), but it is a regression on a documented shape.

**D2 — a URL inside a `note:` segment now enters the URL gate.** `scan.mjs:1198` searches the whole entry body:

```
"- [ ] local:jds/x.md | Acme Corp | Staff Engineer | note: original https://competitor.com/x"
   baseline gate 1 → []
   patched  gate 1 → ["https://competitor.com/x"]
```

A hand-written URL in free text enters the seen-set and the scanner skips that posting silently. Narrow to trigger, but it is the expensive direction, and it is the one I called disqualifying in round 1.

**One scoping closes both, and I ran it before asking.** All six documented forms put the URL in cell 0, or in cell 1 behind a `#NNN` / `#--` / `[NNN](reports/…)` leader. None puts it further out:

- `scan.mjs:1198` → search `entry.body.split('|').slice(0, 2)` and return the first cell match.
- `scan.mjs:1230` → `cells.slice(0, 2).findIndex(...)`, and when that returns -1, fall back to `urlIndex = 0` if `cells[0]` is a non-empty token with no spaces (the old `\S+` semantics, which also excluded `"- [ ] Acme Corp | Staff Engineer | Remote"`).

Applied in a worktree: the six forms stay identical, everything the PR fixes still gets fixed, `local:` returns to baseline behavior, the note exposure closes, and the suite stays at 3668 passed / 0 failed including all your new tests. Two lines and two cases.

Two things not on you: the `applications.md` branch at `scan.mjs:1264` still keeps `)` differently, which you flagged yourself and which is correctly out of scope; and the bot greeted you as a first-time contributor despite #2063, #2070 and #2072 back in July. That greeting bug is fixed in main today.

**Comment by santifer** [comment]
The part that decides this is that you kept the two gates at different widths and said why, @bravely.

Widening the URL match is safe: worst case an entry dedupes that wouldn't have. Widening the company/role gate is not, because one role key suppresses every other posting sharing it — so a marker matched too eagerly there doesn't lose one link, it buries a whole class of live postings. Two rules that look like they should be unified, kept apart deliberately, with the asymmetry written down. That's the note the next person needs and would never guess.

Three more calls I'd have got wrong on my own:

- **`- [!]` excluded from the recognized markers.** A URL that couldn't be fetched isn't evidence a posting was surfaced, so seeding from it would suppress the real one.
- **Strikethrough read at the entry boundary, not line-wide.** `~` is a legal URL character and a `note:` column can carry its own — searching for it globally would truncate real URLs and strip live entries of their pair.
- **Indentation tolerated on the URL gate, rejected on the strict one.** Same asymmetry, applied consistently rather than as a special case.

Merged 🚀
